autonat/autonat-v2.md

```diff
@@ -0,0 +1,207 @@
+`ResponseStatus`. see [Requirements For
+ResponseStatus](#requirements-for-responsestatus)
+
+The initiator MUST check that the nonce received in the `DialAttempt` is the
+same as the nonce the initiator sent in the `Candidate` for the candidate
+index received in `DialResponse`. If the nonce is different, the initiator MUST
+discard this response.
+
+
+### Requirements for ResponseStatus
+
+On receiving a `DialRequest` the peer selects the first candidate on the list it
+is capable of dialing. This candidate address is referred to as _addr_. The
+`ResponseStatus` sent by the peer in the `DialResponse` message MUST be set
+according to the following requirements
+
+`OK`: the peer was able to dial _addr_ successfully.
+
+`E_DIAL_ERROR`: the peer attempted to dial _addr_ and was unable to connect. 
+
+`E_DIAL_REFUSED`: the peer didn't attempt a dial because of rate limiting,
+resource limit reached or blacklisting.
+
+`E_TRANSPORT_NOT_SUPPORTED`: the peer didn't have the ability to dial any of the
+requested addresses.
+
+`E_BAD_REQUEST`: the peer didn't attempt a dial because it was unable to decode
+the message.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+peer that prevented it from completing the request.
+
+Implementations MUST discard responses with status codes they do not understand
+
+### Consideration for DDOS Prevention
+
+In order to prevent attacks like the one described in [RFC 3489, Section
+12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (see excerpt
+below), implementations MUST NOT dial any multiaddress unless it is based on the
+IP address the requesting node is observed as. This restriction as well implies
+that implementations MUST NOT accept dial requests via relayed connections as
+one can not validate the IP address of the requesting node.
+
+> RFC 3489 12.1.1 Attack I: DDOS Against a Target
+>
+> In this case, the attacker provides a large number of clients with the same
+> faked MAPPED-ADDRESS that points to the intended target. This will trick all
+> the STUN clients into thinking that their addresses are equal to that of the
+> target. The clients then hand out that address in order to receive traffic on
+> it (for example, in SIP or H.323 messages). However, all of that traffic
+> becomes focused at the intended target. The attack can provide substantial
+> amplification, especially when used with clients that are using STUN to enable
+> multimedia applications.
+
+
+## Implementation Suggestions
+
+For any given address, implementations SHOULD do the following
+- periodically recheck reachability status
+- query multiple peers to determine reachability
+
+The suggested heuristic for implementations is to consider an address reachable
+if more than 3 peers report a successful dial and to consider an address
+unreachable if more than 3 peers report unsuccessful dials. 
+
+Implementations are free to use different heuristics than this one
+
+
+## RPC Messages
```

Comment:
**Stream behavior**

Do we have consensus that multiple `DialRequest`s should not be pipelined on a single stream? If so, mind adding a paragraph along the lines of:

```
Each `DialRequest` is sent on a new stream. Multiple `DialRequest`s are never sent on the same stream (i.e. no pipelining).
```

Also is a node supposed to close the write side of a stream once it sent the `DialRequest`, `DialResponse` and `AttemptMessage`?

Reply:
No pipelining, please!

I think it's sufficient to define when a stream is closed though, we don't need to explicitly forbid pipelining.

Reply:
I've added lines. 

https://github.com/libp2p/specs/blob/autonat-v2/autonat/autonat-v2.md?plain=1#L57
>Each `DialRequest` is sent on a new stream.


https://github.com/libp2p/specs/blob/autonat-v2/autonat/autonat-v2.md?plain=1#L67-L68
>The peer MUST close this
stream after sending the `DialAttempt` message.


https://github.com/libp2p/specs/blob/autonat-v2/autonat/autonat-v2.md?plain=1#L81-L82
>The peer MUST close the stream after sending the response. The initiator MUST
close the stream after receiving the response.


Does this work?

Reply:
Yes, thank you @sukunrt!